src/lazy/expanded/macro_evaluator.rs

```diff
@@ -10,208 +10,345 @@
+impl<'top, 'data: 'top, D: LazyDecoder<'data>> MacroExpansion<'top, 'data, D> {
     /// Continues evaluating this macro until it:
     ///   * produces another value.
     ///   * encounters another macro or variable that needs to be expanded.
     ///   * is completed.
-    fn next<'top>(
+    fn next(
         &mut self,
         context: EncodingContext<'top>,
-    ) -> IonResult<MacroExpansionStep<'top, 'data, D, M>>
-    where
-        'data: 'top,
-        M: 'top,
-    {
+        environment: Environment<'top, 'data, D>,
+    ) -> IonResult<Option<ValueExpr<'top, 'data, D>>> {
         use MacroExpansionKind::*;
         // Delegate the call to `next()` based on the macro kind.
         match &mut self.kind {
-            MakeString(make_string_expansion) => make_string_expansion.next(context),
-            Values(values_expansion) => values_expansion.next(context),
+            MakeString(make_string_expansion) => make_string_expansion.next(context, environment),
+            Values(values_expansion) => values_expansion.next(context, environment),
             // `void` is trivial and requires no delegation
-            Void => Ok(MacroExpansionStep::Complete),
+            Void => Ok(None),
+            Template(template_expansion) => template_expansion.next(context, environment),
         }
     }
 }
 
-/// Represents a single step in the process of evaluating a macro.
-pub enum MacroExpansionStep<'top, 'data, D: LazyDecoder<'data>, M: MacroInvocation<'data, D>> {
-    /// The next value produced by continuing the macro evaluation.
-    ExpandedValue(LazyExpandedValue<'top, 'data, D>),
-    /// Another macro that will need to be evaluated before an expanded value can be returned.
-    AnotherMacroToEvaluate(M),
-    /// This macro will not produce any further values.
-    Complete,
-}
+pub type MacroStack<'top, 'data, D> = BumpVec<'top, MacroExpansion<'top, 'data, D>>;
+pub type EnvironmentStack<'top, 'data, D> = BumpVec<'top, Environment<'top, 'data, D>>;
 
 /// Evaluates macro invocations recursively, yielding a single expanded value at a time.
 ///
-/// This evaluator can be used in a variety of contexts. It supports the cross product of three
-/// use case dimensions:
+/// The evaluator supports the cross product of three use case dimensions:
 ///
 ///   {e-expression, template macro invocation}
 ///   x {text, binary}
 ///   x {eager, lazy}
-pub struct MacroEvaluator<
-    'data,
-    // The Ion format of the data stream
-    D: LazyDecoder<'data>,
-    // The syntactic element representing the e-expression or template macro invocation
-    M: MacroInvocation<'data, D>,
-    // The storage being used to store the macro expansion stack. (Either a long-lived `Vec` or
-    // a bumpalo [`Bump`](BumpVec) whose contents only live as long as the reader is parked on
-    // the same top-level value).
-    S: Stack<MacroExpansion<'data, D, M>>,
```

Comment:
🗺️ Previously, `MacroEvaluator` was generic over its `Stack` implementation, allowing it to hold either a static `Vec` or a `BumpVec`. This was done so that top level macros could produce multiple values and the `Vec` lived long enough to persist their expansion state between calls to `next()`.

Thanks to a small amount of `unsafe` code, it is now possible for the root `MacroEvaluator` that used to live in the `LazyExpandingReader` to instead live in the bump allocator. The `LazyExpandingReader` holds a raw pointer to it.

This enables the evaluator (and its macro stack) to hold references to other items in the bump allocator, symbols in the symbol table, macros in the macro table, and even other evaluators. They all have the same lifetime, `'top`, which refers to the duration for which the reader is parked on the current top level expression.